rasa/nlu/extractors/entity_synonyms.py

```diff
@@ -1,102 +1,142 @@
 import rasa.utils.io
 import rasa.shared.utils.io
+from rasa.nlu.extractors._entity_synonyms import EntitySynonymMapper
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
 
+# This is a workaround around until we have all components migrated to `GraphComponent`.
+EntitySynonymMapper = EntitySynonymMapper
 
-class EntitySynonymMapper(EntityExtractor):
-    @classmethod
-    def required_components(cls) -> List[Type[Component]]:
-        return [EntityExtractor]
+
+class EntitySynonymMapperComponent(GraphComponent, EntityExtractorMixin):
+    """Maps entities to their synonyms if they appear in the training data."""
+
+    SYNONYM_FILENAME = "synonyms.json"
 
     def __init__(
         self,
-        component_config: Optional[Dict[Text, Any]] = None,
+        config: Optional[Dict[Text, Any]],
+        model_storage: ModelStorage,
+        resource: Resource,
         synonyms: Optional[Dict[Text, Any]] = None,
     ) -> None:
-
-        super().__init__(component_config)
+        """Creates the mapper.
+
+        Args:
+            config: The mapper's config.
+            model_storage: Storage which the component can use to persist and load
+                itself.
+            resource: Resource locator for this component which can be used to persist
+                and load itself from the `model_storage`.
+            synonyms: A dictionary of previously known synonyms.
+        """
+        self._config = config
+        self._model_storage = model_storage
+        self._resource = resource
 
         self.synonyms = synonyms if synonyms else {}
 
-    def train(
-        self,
-        training_data: TrainingData,
-        config: Optional[RasaNLUModelConfig] = None,
-        **kwargs: Any,
-    ) -> None:
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        synonyms: Optional[Dict[Text, Any]] = None,
+    ) -> EntitySynonymMapperComponent:
+        """Creates component (see parent class for full docstring)."""
+        return cls(config, model_storage, resource, synonyms)
 
+    def train(self, training_data: TrainingData,) -> None:
+        """Trains the synonym lookup table."""
         for key, value in list(training_data.entity_synonyms.items()):
-            self.add_entities_if_synonyms(key, value)
+            self._add_entities_if_synonyms(key, value)
 
         for example in training_data.entity_examples:
             for entity in example.get(ENTITIES, []):
                 entity_val = example.get(TEXT)[entity["start"] : entity["end"]]
-                self.add_entities_if_synonyms(entity_val, str(entity.get("value")))
+                self._add_entities_if_synonyms(entity_val, str(entity.get("value")))
+
+        if self._resource:
+            self._persist()
```

Comment:
as the `resource` is not `Optional`
```suggestion
        self._persist()
```